### DISCUSSION TITLE
Export beatmap to v14 file format

### DISCUSSION THREAD

**Discussion by OliBomby** [ppy/osu#18066] [Ideas]
The idea is to add another 'legacy' export button in the editor which exports a beatmap specifically towards osu!stable. This means creating a file format v14 beatmap instead of v128. This would be a lossy kind of conversion because it should truncate time values, round coordinates, and convert sliders with multiple path types into beziers with the same shape. This ensures that the exported beatmap is compatible with osu!stable and bears the most resemblense to the original beatmap.

This idea was suggested to me by people in the Mapping Tools community who wanted a tool for this because they map in lazer for stable. I think having this feature will make the lazer editor more appealing for mappers who currently map in stable and thus bring more mappers to use the lazer editor and provide feedback.

Also I know how to implement all of this myself, even the conversion for sliders to bezier, so if this gets accepted i'll probably go work on a merge request for this.

**Comment by peppy** [comment]
Either this – or adding support for the new features to stable – needs to be done at some point.

Converting sliders to bezier is a definitely DO NOT DO THIS, though. Before you ask, yes I understand it can be done and is being done on ranked beatmaps every day. And it's horrible and should never have been a thing. The constructed data points make life hard.

It will need basic stable support or be a lazer-only feature.

**Reply by OliBomby** [reply]
I do think adding support for these features to stable is a better option but I dont see that happening because stable is feature locked. 

You are overreacting about the badness of converting sliders to bezier. I dont see whose life is made hard by the constructed bezier anchors. It wont happen much that the mapper still edits the sliders after they are made in lazer and if they do its still not that bad to work with bezier. Mappers are already used to making shapes with bezier anyways.

If the multi-type sliders are not converted to bezier, they will look completely wrong in stable.

**Reply by peppy** [reply]
If possible, I would like to avoid this by adding backwards compatibility support to stable for the new curve types (should be very easy, but obviously would not work in the editor). But an argument that can be made is that this would stop third party osu! implementations from working.

If you want to do up a proof-of-concept conversion I'm willing to give it consideration.

**Reply by OliBomby** [reply]
I have a proof-of-concept conversion in C# which is already being used by my Slider Merger tool for years.
https://github.com/OliBomby/Mapping_Tools/blob/master/Mapping_Tools/Classes/ToolHelpers/BezierConverter.cs
It converts perfect circles to bezier by finding the best preset and shortening that preset to the desired circle arc angle using Casteljau's algorithm. The presets were generated with an optimizer beforehand and they are at the maximum arc angle that can be accurately approximated with the amount of bezier anchors they have. Therefore these conversions always use the least number of anchors required for the wanted circle arc angle and they are super fast. I also have exact conversions for catmull to bezier.
Screenshots:
![osu!_dQcGF5Tz43](https://user-images.githubusercontent.com/17460441/166459501-e03463ff-3cb8-45e4-b256-4709533b9485.jpg)
![osu!_GpGcMmEvql](https://user-images.githubusercontent.com/17460441/166459538-f668cebb-eb76-4fd6-b168-99a84edb6308.jpg)
![osu!_OdklskDQcl](https://user-images.githubusercontent.com/17460441/166459560-05e81e01-d342-49ec-8fe0-7361cdabec88.png)
![osu!_OjGdX0hnef](https://user-images.githubusercontent.com/17460441/166459576-91d4ab16-7521-4627-aaa4-2ea40e7cadfa.jpg)

**Reply by peppy** [reply]
Just to make sure it isn't missed, this was implemented in the game via https://github.com/ppy/osu/pull/20952. All that needs to be done is to hook that up to the export process.

**Comment by bdach** [comment]
I'll just add that things like the coordinate truncation already cause issues in lazer itself, e.g. see https://github.com/ppy/osu/issues/17485. So that may get changed period, rather than get changed only for this "compatibility mode".

**Comment by okolenmi** [comment]
Please, add support for stable. New slider features are really good. It's sad to see broken sliders after export.

Before export:
![изображение](https://user-images.githubusercontent.com/55319814/219443186-b11bccb6-8a10-493d-a6bf-e9425333d6ee.png)

After export:
![изображение](https://user-images.githubusercontent.com/55319814/219444503-70597da0-fa52-4389-9f57-a49ecaae229e.png)

**Reply by mk56-spn** [reply]
There's no need for adding to a thread on issues that are known and  are going to be addressed at some point. It's just noise, when a suitable method is found to solve the situation and someone has time to implement it it'll be done

**Reply by okolenmi** [reply]
Support of v128 is important for mappers community. It's not a thing that should be named as "noise". Last message was 9 months ago. Therefore, I wanted to remind developers about it in order to increase the chance of solving the problem in the near future.

**Reply by mk56-spn** [reply]
I never said it was not, simply that bumping threads without providing anything new only serves to  consume dev time, i very much doubt such a fundamental compatibility issue has been forgotten.

**Reply by Axilotl17** [reply]
there is an external tool for this @okolenmi 

https://github.com/JPK314/LazerToStable

use it if you want.

**Reply by Axilotl17** [reply]
to be clear, this will be implemented into the game at some point, but this external tool works for now.
